src/coreclr/inc/check.h

```diff
@@ -16,6 +16,14 @@
 #include "daccess.h"
 #include "unreachable.h"
 
+// Macros for creating type traits to check if a member exists
+#define DEFINE_MEMBER_EXISTENCE_CHECK(Member) \
+template<typename T, typename = void> \
+struct has_##Member : std::false_type {}; \
+\
+template<typename T> \
+struct has_##Member<T, decltype((void)&T::Member, void())> : std::true_type {};
```

Comment:
Can we use the detection idiom here instead? (We have a usage of it already in dacenumerablehash.inl, we can move some of that here.)

Reply:
Could you point me to it? Is it: https://github.com/dotnet/runtime/blob/f0edcd8ac87ff9f7f71d5cc3faf7429c18734b55/src/coreclr/vm/dacenumerablehash.inl#L475

Reply:
I implemented the C++17 version first, and it was quite straightforward:

```
#define EXECUTE_IF_MEMBER_EXISTS(Type, Member, CodeBlock) \
if constexpr (has_##Member<Type>::value) { \
    CodeBlock \
}
```

but then this enable_if(cond), enable_if(!cond) clutter came to support std=c++11. We can definitely simplify some of this code (revert parts of https://github.com/dotnet/runtime/commit/856e8c3510cb0aaaa3b1a1679552c03ae5308b9e) when we are ready to drop support for older compilers. 😅

---

I think the some parts of the product had issues even with clang-9 when we had to switch to clang-12+. Not sure if build with old compiler make sense given those run-time issues; especially when we support the old libc with latest toolchain via cross-compilation? Perhaps it's already the case that https://github.com/dotnet/runtime/blob/1ee61b8667d9db31c0fe53d2e1abc274fb486cfc/eng/common/native/init-compiler.sh#L66-L67
should be:
 ```sh
if [ "$compiler" = "clang" ]; then versions="18 17 16 15 14 13" 
elif [ "$compiler" = "gcc" ]; then versions="14 13 12"; fi
```
for more than one reason and simplify clang<v13 checks (https://github.com/dotnet/runtime/pull/102278/files) even further.